### PR TITLE
Ratelimit support

### PR DESCRIPTION
**Pull request by konrad-kaminski**

### CODE REVIEW

**Review by kukulam-allegro — CHANGES_REQUESTED** [review]

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RateLimitFilterFactoryTest.kt [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+        // then
+        assertThat(filter!!.name).isEqualTo("envoy.lua.ratelimit")
+        assertThat(filter.typedConfig).isInstanceOf(com.google.protobuf.Any::class.java)
+        val config = filter.typedConfig as com.google.protobuf.Any
+        assertThat(config.`is`(Lua::class.java)).isTrue()
+        val lua = config.unpack(Lua::class.java)
+        assertThat(lua.inlineCode.smartTrimmed()).containsExactly(*"""
+            function envoy_on_request(request_handle)
+                request_handle:streamInfo():dynamicMetadata():set("envoy.filters.http.ratelimit.override", "_ecb42b24b3c81841bfdd244fedf430f8", {
```
[the remark this answers is not in the document]
Comment (kukulam-allegro) [reply envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RateLimitFilterFactoryTest.kt:84]: service name before `"_ecb42b24b3c81841bfdd244fedf430f8"`

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/routes/EnvoyIngressRoutesFactory.kt [hunk 2]
```diff
@@ -46,6 +57,57 @@ class EnvoyIngressRoutesFactory(
...
+            .addIngressRateLimits(serviceName, rateLimitEndpoints)
+    }
+
+    private fun RouteAction.Builder.addIngressRateLimits(
+        serviceName: String,
+        rateLimitEndpoints: List<IncomingRateLimitEndpoint>
+    ): RouteAction.Builder = apply {
+        rateLimitEndpoints.forEach { endpoint ->
+            val ruleId = getRuleId(serviceName, endpoint)
+            val hvm = RateLimit.Action.HeaderValueMatch.newBuilder()
```
[the remark this answers is not in the document]
Comment (konrad-kaminski) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/routes/EnvoyIngressRoutesFactory.kt]: Fixed.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/SnapshotUtils.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+fun getRuleId(serviceName: String, endpoint: IncomingRateLimitEndpoint): String {
+    val methods = endpoint.methods.sorted().joinToString()
+    val clients = endpoint.clients.sorted().joinToString(transform = { "${it.name},${it.selector}" })
+    val key = "$serviceName,${endpoint.path},${endpoint.pathMatchingType},$methods,$clients"
+
+    return "${serviceName.replace("-", "_")}_${key.md5()}"
+}
+
+@Suppress("MagicNumber")
+private fun String.md5(): String {
```
[the remark this answers is not in the document]
Comment (konrad-kaminski) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/SnapshotUtils.kt:16]: Yes. The rule id will be a part of every request to ratelimit service and therefore in order to reduce the bandwidth MD5 of the rule is used.

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RateLimitFilterFactoryTest.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (konrad-kaminski) [reply envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RateLimitFilterFactoryTest.kt]: Fixed.

**Review by kukulam-allegro — APPROVED** [review]

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/EnvoySnapshotFactory.kt [hunk 4]
```diff
@@ -213,12 +214,13 @@ class EnvoySnapshotFactory(
...
     private fun getServicesEndpointsForGroup(
+        rateLimitEndpoints: List<IncomingRateLimitEndpoint>,
         globalSnapshot: GlobalSnapshot,
         egressRouteSpecifications: Collection<RouteSpecification>
-    ): List<ClusterLoadAssignment> {
-        return egressRouteSpecifications
-            .mapNotNull { globalSnapshot.endpoints.resources()[it.clusterName] }
-    }
+    ): List<ClusterLoadAssignment> =
+        (egressRouteSpecifications.map { it.clusterName } +
```
Comment (Automaat) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/EnvoySnapshotFactory.kt:221]: this is not readable for me. Too much is happening in this two lines. Can you split it and make it easier to digest?

Comment (konrad-kaminski) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/EnvoySnapshotFactory.kt:221]: Fixed.

**Review by kukulam-allegro — APPROVED** [review]

**Review by Automaat — APPROVED** [review]

**Pull request merged**